sw/device/lib/testing/kmac/kmac_vector_generate.c

```diff
@@ -0,0 +1,67 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "KeccakHash.h"
+#include "SP800-185.h"
+#include "config.h"
+
+/**
+ * Exit if `condition` is false.
+ *
+ * @param condition Condition to be checked.
+ */
+static void assert(int condition) {
+  if (!condition) {
+    printf("Condition not satisfied.\n");
+    exit(1);
+  }
+}
+
+/**
+ * Print given long byte array with hex characters.
+ *
+ * @param ptr Byte array pointer.
+ * @param len Number of bytes to be printed.
+ */
+void print_long_array(uint8_t *ptr, size_t len) {
+  for (size_t i = 0; i < len; ++i)
+    printf("%02x", ptr[i]);
+  printf("\n");
+}
+
+/**
+ * Simple example for XKCP for KMAC test vector generation.
+ */
+int main(void) {
+  unsigned char output[512 / 8];
+  int err_status;
+
+  {
+    const BitSequence *key = (const  BitSequence *)
+            "\x40\x41\x42\x43\x44\x45\x46\x47\x48\x49\x4A\x4B\x4C\x4D\x4E\x4F\x50\x51\x52\x53\x54\x55\x56\x57\x58\x59\x5A\x5B\x5C\x5D\x5E\x5F";
+    const BitSequence *input = (const BitSequence *)"\x00\x01\x02\x03";
+    const BitSequence *customization = (const BitSequence *)"";
+    const BitSequence *expected_output = (const  BitSequence *)
+            "\xE5\x78\x0B\x0D\x3E\xA6\xF7\xD3\xA4\x29\xC5\x70\x6A\xA4\x3A\x00\xFA\xDB\xD7\xD4\x96\x28\x83\x9E\x31\x87\x24\x3F\x45\x6E\xE1\x4E";
```

Comment:
I think these could be strict aliasing violations. A `char*` is not necessarily compatible with `BitSequence*`. (Looks like `BitSequence` is a typedef of `uint8_t`, which is not guaranteed to be a character type.)

I think you might as well declare these as BitSequence arrays in the first place. For example:

```c
const BitSequence key[32] = {0x40, 0x41, 0x42, /* ... */ };
```

The other advantage of this approach is that it retains the array's size. Now, when you pass a pointer and length, you can use `ARRAYSIZE` rather than hardcoding the length.

Reply:
I will be using `uint8_t` arrays here. For the `ARRAYSIZE` macro, the definition comes from [a header in `/sw/device`](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/base/macros.h#L37). Just out of curiosity, since this will be a host code, can I use this header on the host side?

Reply:
[.](https://github.com/lowRISC/opentitan/pull/16906#discussion_r1072790443).